### PR TITLE
feat(consensus): allow configuring block gas limit through `ChainSpec`

### PR DESCRIPTION
**Pull request by tcoratger**
Should solve #3233.

### CODE REVIEW

**Review by rkrasiuk — CHANGES_REQUESTED** [review]
https://github.com/paradigmxyz/reth/issues/3233 is more about being able to set block gas limit through the `ChainSpec` instance, not just having a default method on it.

e.g. user runs `reth node --block-gas-limit 1000000` and everywhere where we currently use a constant, it would instead use the user provided value

**Comment by onbjerg** [comment]
Needs a rebase

**Comment by tcoratger** [comment]
> #3233 is more about being able to set block gas limit through the `ChainSpec` instance, not just having a default method on it.
> 
> e.g. user runs `reth node --block-gas-limit 1000000` and everywhere where we currently use a constant, it would instead use the user provided value

@rkrasiuk I just added `block_gas_limit` to the CLI. I still keep a default value in `ChainSpec` directly because many tests and functions only ask for a default value and do not have access to the `ChainSpec` instance provided by the CLI. For example, there is the following implementation for which no argument is provided and therefore only asks for a default value without fetching additional information as I understand it:
```rust
impl<T: PoolTransaction> Default for AllTransactions<T> {
    fn default() -> Self {
        Self {
            max_account_slots: MAX_ACCOUNT_SLOTS_PER_SENDER,
            minimal_protocol_basefee: MIN_PROTOCOL_BASE_FEE,
            block_gas_limit: ChainSpec::block_gas_limit_default(),
            by_hash: Default::default(),
            txs: Default::default(),
            tx_counter: Default::default(),
            last_seen_block_number: 0,
            last_seen_block_hash: Default::default(),
            pending_basefee: Default::default(),
        }
    }
}
```

File: bin/reth/src/node/mod.rs [hunk 1]
```diff
@@ -141,6 +141,10 @@ pub struct Command {
     /// Automatically mine blocks for new transactions
     #[arg(long)]
     auto_mine: bool,
+
+    /// Configuring block gas limit
+    #[arg(long, default_value = "30000000")]
+    block_gas_limit: u64,
```
Comment (onbjerg) [inline bin/reth/src/node/mod.rs:147]: It does not make sense to expose this as it is network dependent, not something you can just configure. Thus, it should *only* be in the chainspec, not on the CLI. Please remove.

File: bin/reth/src/node/mod.rs [hunk 2]
```diff
@@ -152,8 +156,15 @@ impl Command {
         // Does not do anything on windows.
         raise_fd_limit();
 
+        // Create a new Arc with the modified ChainSpec
+        let chain_spec_with_gas_limit = Arc::new(
+            Arc::try_unwrap(self.chain.clone())
+                .expect("Expected a single reference to ChainSpec")
+                .chainspec_with_block_gas_limit(self.block_gas_limit),
```
Comment (onbjerg) [inline bin/reth/src/node/mod.rs:163]: This is incorrect and it makes the node unable to sync. Refer to CI.

Per above comment, this can just be removed.

**Review by onbjerg — CHANGES_REQUESTED** [review]

File: bin/reth/src/node/mod.rs (hunk 1 above)
Comment (tcoratger) [reply bin/reth/src/node/mod.rs:147]: @onbjerg Yes it seemed strange to me to be able to configure this on the CLI that's why I had made the first version. So this is like going back to commit 6ffe1a6c18e7e6bc5297fe1cdb99950861e636ff where the block gas limit was just exposed as a constant inside `ChainSpec`, correct?

Comment (onbjerg) [reply bin/reth/src/node/mod.rs:147]: Yes, although it should be possible to set it in the actual chainspec JSON of course. So if I made a new chainspec for my own network, I should be able to increase or decrease it :)

Comment (tcoratger) [reply bin/reth/src/node/mod.rs]: @onbjerg Should be ok now. Places where the default value is used remains at the default constant `ETHEREUM_BLOCK_GAS_LIMIT` while the other places directly take into account the `block_gas_limit` item of the `ChainSpec` structure.

**Comment by onbjerg** [comment]
@tcoratger It's still not entirely done, the main issue here is that even if I set a different gas limit in my chainspec, this will not actually affect much of the node. For example:

- All the tracing RPCs still use a gas limit of `30_000_000` because `prepare_call_env` just does `env.tx.gas_limit = ChainSpec::block_gas_limit_default()`, which is incorrect. It should use the spec.
- The transaction pool suffers the same issue, it just uses `ChainSpec::block_gas_limit_default()`, so if I changed the limit, I would still end up building incorrect blocks
- The payload builder also defaults to `30_000_000`, but at least this is configurable via the CLI.

**Comment by tcoratger** [comment]
> @tcoratger It's still not entirely done, the main issue here is that even if I set a different gas limit in my chainspec, this will not actually affect much of the node. For example:
> 
> * All the tracing RPCs still use a gas limit of `30_000_000` because `prepare_call_env` just does `env.tx.gas_limit = ChainSpec::block_gas_limit_default()`, which is incorrect. It should use the spec.
> * The transaction pool suffers the same issue, it just uses `ChainSpec::block_gas_limit_default()`, so if I changed the limit, I would still end up building incorrect blocks
> * The payload builder also defaults to `30_000_000`, but at least this is configurable via the CLI.

@onbjerg Yes I suspected this problem. This will probably sound like a stupid question to you as I'm sure it's obvious with familiarity with the codebase, but in those two functions you mention, I didn't see any variable that could hold the instance of `ChainSpec`, so how to access it?

**Comment by onbjerg** [comment]
@tcoratger if there is not, you'd have to add it, or make the value propagate some other way - follow the code path. Unfortunately I am not super familiar with those 2 parts specifically :)

**Comment by gakonst** [comment]
@mattsse should we just pass in the chainspec as it'd include default block gas limit and it'd be useful across the board in components? and have our builder structs maybe default to it having some value (eg mainnet)

**Pull request closed without merging**